compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -0,0 +1,247 @@
+        Linkage::WeakAny | Linkage::WeakODR => Some(".weak"),
+        Linkage::LinkOnceAny | Linkage::LinkOnceODR => Some(".weak"),
+        Linkage::Internal | Linkage::Private => None, // just doesn't emit any attribute
+        Linkage::ExternalWeak => None,                // terminates with sigill on godbolt
+        Linkage::AvailableExternally => None,         // does not even emit the definition
+        Linkage::Appending => None,                   // only valid on global variables
+        Linkage::Common => None,                      // function may not have common linkage
+    }
+}
+
+fn prefix_and_suffix<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    instance: Instance<'tcx>,
+    item_data: &MonoItemData,
+) -> (String, String) {
+    use std::fmt::Write;
+
+    let target = &tcx.sess.target;
+    let target_arch = tcx.sess.asm_arch;
+
+    let is_arm = target.arch == "arm";
+    let is_thumb = is_arm && target.llvm_target.contains("thumb");
+
+    let mangle = (target.is_like_windows && matches!(target_arch, Some(InlineAsmArch::X86)))
+        || target.options.vendor == "apple";
+
+    let asm_name = format!("{}{}", if mangle { "_" } else { "" }, tcx.symbol_name(instance).name);
+
+    let opt_section = tcx
+        .get_attr(instance.def.def_id(), sym::link_section)
+        .and_then(|attr| attr.value_str())
+        .map(|attr| attr.as_str().to_string());
+
+    let attrs = tcx.codegen_fn_attrs(instance.def_id());
+    let (arch_prefix, arch_suffix) = if is_arm {
+        (
+            match attrs.instruction_set {
+                None => match is_thumb {
+                    true => ".thumb\n.thumb_func",
+                    false => ".arm",
+                },
+                Some(InstructionSetAttr::ArmA32) => ".arm",
+                Some(InstructionSetAttr::ArmT32) => ".thumb\n.thumb_func",
+            },
+            match is_thumb {
+                true => ".thumb",
+                false => ".arm",
+            },
+        )
+    } else {
+        ("", "")
+    };
+
+    let mut begin = String::new();
+    let mut end = String::new();
+    match AsmBinaryFormat::from_target(&tcx.sess.target) {
+        AsmBinaryFormat::Elf => {
+            let section = opt_section.unwrap_or(format!(".text.{asm_name}"));
+
+            let progbits = match is_arm {
+                true => "%progbits",
+                false => "@progbits",
+            };
+
+            let function = match is_arm {
+                true => "%function",
+                false => "@function",
+            };
+
+            writeln!(begin, ".pushsection {section},\"ax\", {progbits}").unwrap();
+            writeln!(begin, ".balign 4").unwrap();
+            if let Some(linkage) = linkage_directive(item_data.linkage) {
+                writeln!(begin, "{linkage} {asm_name}").unwrap();
+            }
+            if let Visibility::Hidden = item_data.visibility {
+                writeln!(begin, ".hidden {asm_name}").unwrap();
+            }
```

Comment:
I've not been able to actually generate code that triggers this if. Visibility just seems to always be `Default`. So this is entirely untested at the moment.

Reply:
For anything like this, drop a question on Zulip https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp. Unfortunately that enum doesn't seem to be very well documented